File: src/gpbiometricspy/data_io_cleaning.py
from __future__ import annotations

from collections import OrderedDict
from pathlib import Path
import re
import warnings

import numpy as np
import pandas as pd


class GazepointSessionData(OrderedDict):
    """Named mapping of Gazepoint session exports with R-style metadata."""

    def __init__(self, *args, dir=None, session=None, file_index=None, **kwargs):
        super().__init__(*args, **kwargs)
        self.dir = dir
        self.session = session
        self.file_index = file_index
        self.attrs = {
            "dir": dir,
            "session": session,
            "file_index": file_index,
            "class": ["gazepoint_session_data", "list"],
        }


def _clean_path(value) -> str:
    return Path(value).resolve(strict=False).as_posix()


def _guess_delimiter(path: Path) -> str:
    try:
        first = path.read_text(encoding="utf-8-sig", errors="replace").splitlines()[0]
    except (IndexError, OSError):
        return ","
    counts = {",": first.count(","), ";": first.count(";"), "\t": first.count("\t")}
    # R which.max() selects the first maximum in comma, semicolon, tab order.
    return max(counts, key=counts.get) if any(counts.values()) else ","


def _detect_type(path: Path) -> str:
    name = path.name.lower()
    if re.search(r"all[_ -]?gaze|allgaze", name):
        return "all_gaze"
    if re.search(r"fixation|fixations", name):
        return "fixations"
    if "summary" in name:
        return "summary"
    if re.search(r"biometric|gsr|eda|ppg|bvp|heart|hr|ibi|rri", name):
        return "biometrics"
    if re.search(r"marker|trigger|ttl|event", name):
        return "markers"
    return "unknown"


def _safe_name(path: Path, session=None) -> str:
    name = path.stem
    sessions = [] if session is None else ([session] if isinstance(session, str) else list(session))
    for ss in sessions:
        name = re.sub(rf"^{re.escape(str(ss))}([_ -]+)?", "", name, flags=re.I)
    name = re.sub(r"[^A-Za-z0-9_]+", "_", name)
    name = re.sub(r"_+", "_", name).strip("_")
    return name or "gazepoint_file"


def _make_unique(names):
    used = set()
    out = []
    for raw in names:
        candidate = raw
        if candidate in used:
            i = 1
            while f"{raw}_{i}" in used:
                i += 1
            candidate = f"{raw}_{i}"
        used.add(candidate)
        out.append(candidate)
    return out


def _session_keep(files, session=None, session_match="prefix"):
    if session_match not in {"prefix", "contains", "regex"}:
        raise ValueError("`session_match` must be one of: prefix, contains, regex.")
    sessions = [] if session is None else ([session] if isinstance(session, str) else list(session))
    sessions = [str(x) for x in sessions if str(x)]
    if not sessions:
        return [True] * len(files)
    result = []
    for path in files:
        base = path.name
        keep = False
        for ss in sessions:
            if session_match == "regex":
                keep |= re.search(ss, base, flags=re.I) is not None
            elif session_match == "prefix":
                keep |= re.search(rf"^{re.escape(ss)}", base, flags=re.I) is not None
            else:
                keep |= re.search(re.escape(ss), base, flags=re.I) is not None
        result.append(keep)
    return result


def _read_csv(path: Path, file_encoding="UTF-8-BOM"):
    sep = _guess_delimiter(path)
    encoding = "utf-8-sig" if str(file_encoding).upper() == "UTF-8-BOM" else file_encoding
    return pd.read_csv(
        path,
        sep=sep,
        encoding=encoding,
        na_values=["", "NA", "NaN", "N/A", "NULL", "null"],
        keep_default_na=True,
    )


def import_gazepoint_data(
    dir,
    session=None,
    pattern=r"\.csv$",
    recursive=False,
    session_match="prefix",
    file_encoding="UTF-8-BOM",
    add_file_info=True,
):
    """Import Gazepoint-style exports from a session directory.

    This is a direct Python counterpart of the R 2.0.0 session-folder importer.
    The returned ordered mapping exposes ``file_index`` and related metadata as
    attributes, analogous to the R ``gazepoint_session_data`` list attributes.
    """
    if dir is None or not str(dir):
        raise ValueError("Supply `dir`, the folder containing Gazepoint export files.")
    root = Path(dir)
    if not root.is_dir():
        raise FileNotFoundError(f"Folder does not exist: {dir}")
    try:
        regex = re.compile(pattern, flags=re.I)
    except re.error as exc:
        raise ValueError(f"Invalid `pattern`: {exc}") from exc
    iterator = root.rglob("*") if recursive else root.glob("*")
    files = sorted(p for p in iterator if p.is_file() and regex.search(p.name))
    if not files:
        raise ValueError(f"No files matching `pattern` were found in: {dir}")
    keep = _session_keep(files, session=session, session_match=session_match)
    files = [p for p, flag in zip(files, keep) if flag]
    if not files:
        raise ValueError(f"No files matched `session` in: {dir}")

    names = _make_unique([_safe_name(p, session=session) for p in files])
    rows = []
    result = GazepointSessionData(dir=_clean_path(root), session=session)
    for i, (name, path) in enumerate(zip(names, files), start=1):
        dat = _read_csv(path, file_encoding=file_encoding)
        if add_file_info:
            dat["gp_source_file"] = _clean_path(path)
            dat["gp_source_basename"] = path.name
            dat["gp_source_index"] = i
        result[name] = dat
        rows.append(
            {
                "element": name,
                "file": _clean_path(path),
                "basename": path.name,
                "detected_type": _detect_type(path),
                "rows": int(len(dat)),
                "columns": int(dat.shape[1]),
            }
        )
    index = pd.DataFrame(rows)
    result.file_index = index
    result.attrs["file_index"] = index
    return result


def _allowed_missing(missing: np.ndarray, max_gap):
    missing = np.asarray(missing, dtype=bool)
    if not missing.any():
        return np.zeros(len(missing), dtype=bool)
    if np.isinf(max_gap):
        return missing.copy()
    allowed = np.zeros(len(missing), dtype=bool)
    start = None
    for i, flag in enumerate(np.r_[missing, False]):
        if flag and start is None:
            start = i
        elif not flag and start is not None:
            if i - start <= int(max_gap):
                allowed[start:i] = True
            start = None
    return allowed


def _locf(values):
    out = np.asarray(values, dtype=float).copy()
    last = np.nan
    for i in range(len(out)):
        if np.isnan(out[i]):
            out[i] = last
        else:
            last = out[i]
    return out


def _nocb(values):
    return _locf(np.asarray(values, dtype=float)[::-1])[::-1]


def _impute_vector(
    values,
    time=None,
    method="linear",
    max_gap=np.inf,
    fill_edges=True,
    constant_value=0,
    treat_infinite_as_missing=True,
):
    original = np.asarray(values)
    try:
        x = original.astype(float, copy=True)
    except (TypeError, ValueError):
        x = pd.to_numeric(pd.Series(original), errors="coerce").to_numpy(float)
    if time is None:
        tt = np.arange(1, len(x) + 1, dtype=float)
    else:
        tt = pd.to_numeric(pd.Series(time), errors="coerce").to_numpy(float)
        if len(tt) != len(x) or not np.isfinite(tt).all():
            tt = np.arange(1, len(x) + 1, dtype=float)
    missing = np.isnan(x)
    if treat_infinite_as_missing:
        missing |= ~np.isfinite(x)
    x[missing] = np.nan
    allowed = _allowed_missing(missing, max_gap)
    disallowed = missing & ~allowed
    if not allowed.any():
        return x, np.zeros(len(x), dtype=bool)
    observed = ~np.isnan(x)
    if not observed.any():
        if method == "constant":
            out = x.copy()
            out[allowed] = constant_value
            return out, allowed
        warnings.warn("Cannot impute a signal with no observed values.", RuntimeWarning, stacklevel=2)
        return x, np.zeros(len(x), dtype=bool)

    if method == "constant":
        out = x.copy()
        out[allowed] = constant_value
    elif method == "linear":
        out = x.copy()
        if observed.sum() == 1:
            fitted = np.full(len(x), x[observed][0], dtype=float)
        else:
            fitted = np.interp(tt, tt[observed], x[observed])
            if not fill_edges:
                fitted[tt < np.min(tt[observed])] = np.nan
                fitted[tt > np.max(tt[observed])] = np.nan
        out[allowed] = fitted[allowed]
    elif method == "locf":
        fitted = _locf(x)
        if fill_edges:
            fitted = _nocb(fitted)
        out = x.copy()
        out[allowed] = fitted[allowed]
    elif method == "nocb":
        fitted = _nocb(x)
        if fill_edges:
            fitted = _locf(fitted)
        out = x.copy()
        out[allowed] = fitted[allowed]
    elif method == "nearest":
        out = x.copy()
        observed_idx = np.flatnonzero(observed)
        for i in np.flatnonzero(allowed):
            lefts = observed_idx[observed_idx < i]
            rights = observed_idx[observed_idx > i]
            left = lefts[-1] if len(lefts) else None
            right = rights[0] if len(rights) else None
            if left is None and right is None:
                out[i] = np.nan
            elif left is None:
                out[i] = x[right]
            elif right is None:
                out[i] = x[left]
            else:
                out[i] = x[left] if abs(tt[i] - tt[left]) <= abs(tt[right] - tt[i]) else x[right]
    else:
        raise ValueError("`method` must be one of: linear, locf, nocb, nearest, constant.")
    out[disallowed] = np.nan
    # R flags only originally-NA values; Inf may be replaced but is not counted
    # by `is.na(original)`. Preserve that subtle behavior.
    original_na = pd.isna(pd.Series(original)).to_numpy(bool)
    return out, allowed & original_na


def impute_gazepoint_missing(
    data,
    method="linear",
    cols=None,
    time_col=None,
    group_cols=None,
    max_gap=np.inf,
    fill_edges=True,
    constant_value=0,
    add_flags=True,
    treat_infinite_as_missing=True,
):
    """Impute missing Gazepoint signal values with R 2.0.0 semantics."""
    methods = {"linear", "locf", "nocb", "nearest", "constant"}
    if method not in methods:
        raise ValueError("`method` must be one of: linear, locf, nocb, nearest, constant.")
    try:
        gap = float(max_gap)
    except (TypeError, ValueError) as exc:
        raise ValueError("`max_gap` must be a non-negative number or Inf.") from exc
    if np.isnan(gap) or gap < 0:
        raise ValueError("`max_gap` must be a non-negative number or Inf.")
    max_gap = np.inf if np.isinf(gap) else int(gap)

    if isinstance(data, pd.Series) and pd.api.types.is_numeric_dtype(data):
        values, _ = _impute_vector(data.to_numpy(), method=method, max_gap=max_gap,
                                   fill_edges=fill_edges, constant_value=constant_value,
                                   treat_infinite_as_missing=treat_infinite_as_missing)
        return pd.Series(values, index=data.index, name=data.name)
    if isinstance(data, (list, tuple, np.ndarray)):
        arr = np.asarray(data)
        if arr.ndim != 1:
            raise TypeError("`data` must be a numeric vector or data frame.")
        values, _ = _impute_vector(arr, method=method, max_gap=max_gap,
                                   fill_edges=fill_edges, constant_value=constant_value,
                                   treat_infinite_as_missing=treat_infinite_as_missing)
        return values
    if not isinstance(data, pd.DataFrame):
        raise TypeError("`data` must be a numeric vector or data frame.")

    out = data.copy()
    groups = [] if group_cols is None else ([group_cols] if isinstance(group_cols, str) else list(group_cols))
    if groups:
        missing = [c for c in groups if c not in out.columns]
        if missing:
            raise ValueError("Missing grouping columns: " + ", ".join(missing))
    if time_col is not None and time_col not in out.columns:
        raise ValueError("`time_col` not found in `data`.")
    if cols is None:
        cols = [c for c in out.columns if pd.api.types.is_numeric_dtype(out[c]) and c not in set(groups + ([time_col] if time_col else []))]
    elif isinstance(cols, str):
        cols = [cols]
    else:
        cols = list(cols)
    if not cols:
        raise ValueError("No columns selected for imputation.")
    missing = [c for c in cols if c not in out.columns]
    if missing:
        raise ValueError("Missing columns: " + ", ".join(missing))
    non_numeric = [c for c in cols if not pd.api.types.is_numeric_dtype(out[c])]
    if non_numeric:
        raise ValueError("Selected columns must be numeric: " + ", ".join(non_numeric))

    if groups:
        group_indices = [np.asarray(idx, dtype=int) for idx in out.groupby(groups, sort=True, dropna=True).indices.values()]
    else:
        group_indices = [np.arange(len(out), dtype=int)]
    summaries = []
    for col in cols:
        flag = np.zeros(len(out), dtype=bool)
        numeric = pd.to_numeric(out[col], errors="coerce").to_numpy(float)
        before_mask = np.isnan(numeric) | ((~np.isfinite(numeric)) if treat_infinite_as_missing else False)
        before = int(np.sum(before_mask))
        for idx in group_indices:
            tt = out.iloc[idx][time_col].to_numpy() if time_col is not None else None
            values, imputed = _impute_vector(
                out.iloc[idx][col].to_numpy(), time=tt, method=method, max_gap=max_gap,
                fill_edges=fill_edges, constant_value=constant_value,
                treat_infinite_as_missing=treat_infinite_as_missing,
            )
            col_loc = out.columns.get_loc(col)
            out.iloc[idx, col_loc] = values
            flag[idx] = imputed
        numeric_after = pd.to_numeric(out[col], errors="coerce").to_numpy(float)
        after_mask = np.isnan(numeric_after) | ((~np.isfinite(numeric_after)) if treat_infinite_as_missing else False)
        if add_flags:
            out[f"{col}_was_imputed"] = flag
        summaries.append({
            "column": col,
            "n_missing_before": before,
            "n_imputed": int(flag.sum()),
            "n_missing_after": int(after_mask.sum()),
            "method": method,
            "max_gap": np.inf if np.isinf(max_gap) else max_gap,
        })
    out.attrs.update(data.attrs)
    out.attrs["imputation_summary"] = pd.DataFrame(summaries)
    return out

File: reference/R/data-io-cleaning.R

# Gazepoint data import and generic missing-value imputation helpers

.gp_io_clean_path <- function(x) {
  normalizePath(x, winslash = "/", mustWork = FALSE)
}

.gp_io_guess_delimiter <- function(path) {
  first <- readLines(path, n = 1L, warn = FALSE)

  if (!length(first)) {
    return(",")
  }

  counts <- c(
    comma = lengths(regmatches(first, gregexpr(",", first, fixed = TRUE))),
    semicolon = lengths(regmatches(first, gregexpr(";", first, fixed = TRUE))),
    tab = lengths(regmatches(first, gregexpr("\t", first, fixed = TRUE)))
  )

  if (all(counts == 0)) {
    return(",")
  }

  c(comma = ",", semicolon = ";", tab = "\t")[[which.max(counts)]]
}

.gp_io_detect_type <- function(path) {
  b <- tolower(basename(path))

  if (grepl("all[_ -]?gaze|allgaze", b)) return("all_gaze")
  if (grepl("fixation|fixations", b)) return("fixations")
  if (grepl("summary", b)) return("summary")
  if (grepl("biometric|gsr|eda|ppg|bvp|heart|hr|ibi|rri", b)) return("biometrics")
  if (grepl("marker|trigger|ttl|event", b)) return("markers")

  "unknown"
}

.gp_io_safe_name <- function(path, session = NULL) {
  nm <- tools::file_path_sans_ext(basename(path))

  if (!is.null(session) && length(session)) {
    for (ss in session) {
      ss_escaped <- gsub("([][{}()+*^$|\\\\?.])", "\\\\\\1", ss)
      nm <- sub(paste0("^", ss_escaped, "([_ -]+)?"), "", nm, ignore.case = TRUE)
    }
  }

  nm <- gsub("[^A-Za-z0-9_]+", "_", nm)
  nm <- gsub("_+", "_", nm)
  nm <- gsub("^_|_$", "", nm)

  if (!nzchar(nm)) {
    nm <- "gazepoint_file"
  }

  nm
}

.gp_io_session_keep <- function(files,
                                session = NULL,
                                session_match = c("prefix", "contains", "regex")) {
  session_match <- match.arg(session_match)

  if (is.null(session) || !length(session) || all(!nzchar(session))) {
    return(rep(TRUE, length(files)))
  }

  b <- basename(files)

  keep_one <- function(ss) {
    if (session_match == "regex") {
      return(grepl(ss, b, ignore.case = TRUE))
    }

    ss_escaped <- gsub("([][{}()+*^$|\\\\?.])", "\\\\\\1", ss)

    if (session_match == "prefix") {
      return(grepl(paste0("^", ss_escaped), b, ignore.case = TRUE))
    }

    grepl(ss_escaped, b, ignore.case = TRUE)
  }

  Reduce(`|`, lapply(session, keep_one))
}

.gp_io_read_csv <- function(path, file_encoding = "UTF-8-BOM") {
  sep <- .gp_io_guess_delimiter(path)

  utils::read.table(
    file = path,
    header = TRUE,
    sep = sep,
    quote = "\"",
    dec = ".",
    fill = TRUE,
    comment.char = "",
    stringsAsFactors = FALSE,
    check.names = FALSE,
    na.strings = c("", "NA", "NaN", "N/A", "NULL", "null"),
    fileEncoding = file_encoding
  )
}

.gp_impute_find_allowed <- function(is_missing, max_gap) {
  if (!any(is_missing)) {
    return(rep(FALSE, length(is_missing)))
  }

  if (is.infinite(max_gap)) {
    return(is_missing)
  }

  r <- rle(is_missing)
  ends <- cumsum(r$lengths)
  starts <- ends - r$lengths + 1L
  allowed <- rep(FALSE, length(is_missing))

  for (i in seq_along(r$values)) {
    if (isTRUE(r$values[i]) && r$lengths[i] <= max_gap) {
      allowed[starts[i]:ends[i]] <- TRUE
    }
  }

  allowed
}

.gp_impute_locf <- function(x) {
  out <- x
  last <- NA_real_

  for (i in seq_along(out)) {
    if (is.na(out[i])) {
      out[i] <- last
    } else {
      last <- out[i]
    }
  }

  out
}

.gp_impute_nocb <- function(x) {
  rev(.gp_impute_locf(rev(x)))
}

.gp_impute_vector <- function(x,
                              time = NULL,
                              method = c("linear", "locf", "nocb", "nearest", "constant"),
                              max_gap = Inf,
                              fill_edges = TRUE,
                              constant_value = 0,
                              treat_infinite_as_missing = TRUE) {
  method <- match.arg(method)

  original <- x

  if (!is.numeric(x)) {
    x <- suppressWarnings(as.numeric(x))
  }

  if (is.null(time)) {
    time <- seq_along(x)
  } else {
    time <- suppressWarnings(as.numeric(time))
    if (length(time) != length(x) || any(!is.finite(time))) {
      time <- seq_along(x)
    }
  }

  miss <- is.na(x)

  if (isTRUE(treat_infinite_as_missing)) {
    miss <- miss | !is.finite(x)
  }

  x[miss] <- NA_real_

  allowed <- .gp_impute_find_allowed(miss, max_gap = max_gap)
  disallowed <- miss & !allowed

  if (!any(allowed)) {
    return(list(values = x, imputed = rep(FALSE, length(x))))
  }

  observed <- !is.na(x)

  if (!any(observed)) {
    if (method == "constant") {
      out <- x
      out[allowed] <- constant_value
      return(list(values = out, imputed = allowed))
    }

    warning("Cannot impute a signal with no observed values.", call. = FALSE)
    return(list(values = x, imputed = rep(FALSE, length(x))))
  }

  if (method == "constant") {
    out <- x
    out[allowed] <- constant_value
  } else if (method == "linear") {
    if (sum(observed) == 1L) {
      out <- x
      out[allowed] <- x[observed][1L]
    } else {
      rule <- if (isTRUE(fill_edges)) 2 else 1
      fitted <- stats::approx(
        x = time[observed],
        y = x[observed],
        xout = time,
        rule = rule,
        ties = "ordered"
      )$y

      out <- x
      out[allowed] <- fitted[allowed]
    }
  } else if (method == "locf") {
    fitted <- .gp_impute_locf(x)

    if (isTRUE(fill_edges)) {
      fitted <- .gp_impute_nocb(fitted)
    }

    out <- x
    out[allowed] <- fitted[allowed]
  } else if (method == "nocb") {
    fitted <- .gp_impute_nocb(x)

    if (isTRUE(fill_edges)) {
      fitted <- .gp_impute_locf(fitted)
    }

    out <- x
    out[allowed] <- fitted[allowed]
  } else {
    f1 <- .gp_impute_locf(x)
    f2 <- .gp_impute_nocb(x)

    out <- x

    for (i in which(allowed)) {
      left <- max(which(!is.na(x) & seq_along(x) < i), na.rm = TRUE)
      right <- min(which(!is.na(x) & seq_along(x) > i), na.rm = TRUE)

      if (!is.finite(left)) left <- NA_integer_
      if (!is.finite(right)) right <- NA_integer_

      if (is.na(left) && is.na(right)) {
        out[i] <- NA_real_
      } else if (is.na(left)) {
        out[i] <- x[right]
      } else if (is.na(right)) {
        out[i] <- x[left]
      } else {
        out[i] <- if (abs(time[i] - time[left]) <= abs(time[right] - time[i])) x[left] else x[right]
      }
    }
  }

  out[disallowed] <- NA_real_

  list(values = out, imputed = allowed & is.na(original))
}

.gp_impute_group_indices <- function(data, group_cols = NULL) {
  if (is.null(group_cols) || !length(group_cols)) {
    return(list(all = seq_len(nrow(data))))
  }

  missing <- setdiff(group_cols, names(data))

  if (length(missing)) {
    stop("Missing grouping columns: ", paste(missing, collapse = ", "), call. = FALSE)
  }

  split(seq_len(nrow(data)), interaction(data[group_cols], drop = TRUE, sep = " | "))
}

#' Import Gazepoint export files from a session folder
#'
#' Loads Gazepoint-style CSV exports from a folder into a named list of data
#' frames. The function is intended as a single entry point for common session
#' folders that contain all-gaze, fixation, summary, biometric, marker, or
#' related exports.
#'
#' @param dir Folder containing Gazepoint export files.
#' @param session Optional session or participant prefix used to select files.
#'   Can be a character vector.
#' @param pattern File pattern. Defaults to CSV files.
#' @param recursive If TRUE, search subfolders.
#' @param session_match How `session` should be matched: `"prefix"`,
#'   `"contains"`, or `"regex"`.
#' @param file_encoding Encoding passed to `read.table()`.
#' @param add_file_info If TRUE, add source-file columns to each imported data
#'   frame.
#'
#' @return A named list of data frames with class `gazepoint_session_data`.
#'   File metadata are stored in the `file_index` attribute.
#' @export
#'
#' @examples
#' \dontrun{
#' session_data <- import_gazepoint_data("path/to/session", session = "P01")
#' names(session_data)
#' attr(session_data, "file_index")
#' }
import_gazepoint_data <- function(dir,
                                  session = NULL,
                                  pattern = "\\.csv$",
                                  recursive = FALSE,
                                  session_match = c("prefix", "contains", "regex"),
                                  file_encoding = "UTF-8-BOM",
                                  add_file_info = TRUE) {
  session_match <- match.arg(session_match)

  if (missing(dir) || !nzchar(dir)) {
    stop("Supply `dir`, the folder containing Gazepoint export files.", call. = FALSE)
  }

  if (!dir.exists(dir)) {
    stop("Folder does not exist: ", dir, call. = FALSE)
  }

  files <- list.files(
    path = dir,
    pattern = pattern,
    full.names = TRUE,
    recursive = recursive,
    ignore.case = TRUE
  )

  files <- files[file.exists(files)]

  if (!length(files)) {
    stop("No files matching `pattern` were found in: ", dir, call. = FALSE)
  }

  keep <- .gp_io_session_keep(
    files = files,
    session = session,
    session_match = session_match
  )

  files <- files[keep]

  if (!length(files)) {
    stop("No files matched `session` in: ", dir, call. = FALSE)
  }

  names_out <- vapply(files, .gp_io_safe_name, character(1), session = session)
  names_out <- make.unique(names_out, sep = "_")

  out <- vector("list", length(files))
  file_index <- data.frame(
    element = names_out,
    file = .gp_io_clean_path(files),
    basename = basename(files),
    detected_type = vapply(files, .gp_io_detect_type, character(1)),
    rows = NA_integer_,
    columns = NA_integer_,
    stringsAsFactors = FALSE
  )

  for (i in seq_along(files)) {
    dat <- .gp_io_read_csv(files[i], file_encoding = file_encoding)

    if (isTRUE(add_file_info)) {
      dat$gp_source_file <- .gp_io_clean_path(files[i])
      dat$gp_source_basename <- basename(files[i])
      dat$gp_source_index <- i
    }

    out[[i]] <- dat
    file_index$rows[i] <- nrow(dat)
    file_index$columns[i] <- ncol(dat)
  }

  names(out) <- names_out

  attr(out, "dir") <- .gp_io_clean_path(dir)
  attr(out, "session") <- session
  attr(out, "file_index") <- file_index

  class(out) <- c("gazepoint_session_data", "list")

  out
}

#' Impute missing values in Gazepoint signals
#'
#' Interpolates missing values in numeric Gazepoint time series, such as pupil,
#' GSR/EDA, PPG/BVP, heart-rate, IBI/RRI, or other continuous channels. The
#' function can work on a numeric vector, a time-series object, or selected
#' numeric columns of a data frame.
#'
#' @param data Numeric vector, time-series object, or data frame.
#' @param method Imputation method: `"linear"`, `"locf"`, `"nocb"`,
#'   `"nearest"`, or `"constant"`.
#' @param cols Columns to impute when `data` is a data frame. If NULL, all
#'   numeric columns except time and grouping columns are used.
#' @param time_col Optional time column for interpolation.
#' @param group_cols Optional grouping columns. Imputation is performed within
#'   groups.
#' @param max_gap Maximum missing-gap length, in samples, to impute. Longer
#'   gaps remain missing. Defaults to `Inf`.
#' @param fill_edges If TRUE, leading and trailing gaps are filled using the
#'   nearest observed value for methods that support it.
#' @param constant_value Value used when `method = "constant"`.
#' @param add_flags If TRUE and `data` is a data frame, add logical
#'   `<column>_was_imputed` columns.
#' @param treat_infinite_as_missing If TRUE, infinite values are treated as
#'   missing before imputation.
#'
#' @return Object of the same basic type as `data`. Data-frame outputs include
#'   an `imputation_summary` attribute.
#' @export
#'
#' @examples
#' x <- c(1, NA, 3, 4)
#' impute_gazepoint_missing(x)
#'
#' dat <- data.frame(time_s = 1:5, GSR = c(1, NA, 3, NA, 5))
#' impute_gazepoint_missing(dat, cols = "GSR", time_col = "time_s")
impute_gazepoint_missing <- function(data,
                                     method = c("linear", "locf", "nocb", "nearest", "constant"),
                                     cols = NULL,
                                     time_col = NULL,
                                     group_cols = NULL,
                                     max_gap = Inf,
                                     fill_edges = TRUE,
                                     constant_value = 0,
                                     add_flags = TRUE,
                                     treat_infinite_as_missing = TRUE) {
  method <- match.arg(method)

  if (!is.finite(max_gap) && !is.infinite(max_gap)) {
    stop("`max_gap` must be a non-negative number or Inf.", call. = FALSE)
  }

  if (max_gap < 0) {
    stop("`max_gap` must be non-negative.", call. = FALSE)
  }

  max_gap <- if (is.infinite(max_gap)) {
    Inf
  } else {
    as.integer(max_gap)
  }

  if (stats::is.ts(data)) {
    tsp_data <- stats::tsp(data)
    res <- .gp_impute_vector(
      x = as.numeric(data),
      method = method,
      max_gap = max_gap,
      fill_edges = fill_edges,
      constant_value = constant_value,
      treat_infinite_as_missing = treat_infinite_as_missing
    )

    return(stats::ts(res$values, start = tsp_data[1L], end = tsp_data[2L], frequency = tsp_data[3L]))
  }

  if (is.numeric(data) && is.null(dim(data))) {
    res <- .gp_impute_vector(
      x = data,
      method = method,
      max_gap = max_gap,
      fill_edges = fill_edges,
      constant_value = constant_value,
      treat_infinite_as_missing = treat_infinite_as_missing
    )

    return(res$values)
  }

  if (!is.data.frame(data)) {
    stop("`data` must be a numeric vector, ts object, or data frame.", call. = FALSE)
  }

  out <- data

  if (!is.null(time_col) && !time_col %in% names(out)) {
    stop("`time_col` not found in `data`.", call. = FALSE)
  }

  if (is.null(cols)) {
    numeric_cols <- names(out)[vapply(out, is.numeric, logical(1))]
    cols <- setdiff(numeric_cols, c(time_col, group_cols))
  }

  if (!length(cols)) {
    stop("No columns selected for imputation.", call. = FALSE)
  }

  missing_cols <- setdiff(cols, names(out))

  if (length(missing_cols)) {
    stop("Missing columns: ", paste(missing_cols, collapse = ", "), call. = FALSE)
  }

  non_numeric <- cols[!vapply(out[cols], is.numeric, logical(1))]

  if (length(non_numeric)) {
    stop("Selected columns must be numeric: ", paste(non_numeric, collapse = ", "), call. = FALSE)
  }

  groups <- .gp_impute_group_indices(out, group_cols = group_cols)

  summaries <- list()

  for (cc in cols) {
    flag <- rep(FALSE, nrow(out))
    before <- sum(is.na(out[[cc]]) | if (isTRUE(treat_infinite_as_missing)) !is.finite(out[[cc]]) else FALSE)

    for (g in groups) {
      tt <- if (!is.null(time_col)) out[[time_col]][g] else NULL

      res <- .gp_impute_vector(
        x = out[[cc]][g],
        time = tt,
        method = method,
        max_gap = max_gap,
        fill_edges = fill_edges,
        constant_value = constant_value,
        treat_infinite_as_missing = treat_infinite_as_missing
      )

      out[[cc]][g] <- res$values
      flag[g] <- res$imputed
    }

    after <- sum(is.na(out[[cc]]) | if (isTRUE(treat_infinite_as_missing)) !is.finite(out[[cc]]) else FALSE)

    if (isTRUE(add_flags)) {
      out[[paste0(cc, "_was_imputed")]] <- flag
    }

    summaries[[cc]] <- data.frame(
      column = cc,
      n_missing_before = before,
      n_imputed = sum(flag, na.rm = TRUE),
      n_missing_after = after,
      method = method,
      max_gap = if (is.infinite(max_gap)) Inf else max_gap,
      stringsAsFactors = FALSE
    )
  }

  summary <- do.call(rbind, summaries)
  row.names(summary) <- NULL

  attr(out, "imputation_summary") <- summary

  out
}

